File: src/csv_log.py
"""
CSV logging of raw samples captured during a test run.

Every test invocation (baseline or energy capture) writes its raw
per-sample data to a timestamped CSV in the user cache directory, so runs
can be reviewed/replotted later outside the TUI.

Cache location follows XDG conventions: $XDG_CACHE_HOME/jetson-energy-usage
if set, else ~/.cache/jetson-energy-usage.
"""
from __future__ import annotations

import csv
import os
import time
from pathlib import Path

from sampler import Sample


def cache_dir() -> Path:
    base = os.environ.get("XDG_CACHE_HOME")
    root = Path(base) if base else Path.home() / ".cache"
    d = root / "jetson-energy-usage"
    d.mkdir(parents=True, exist_ok=True)
    return d


def write_samples_csv(test_name: str, samples: list[Sample], bus_voltage_v: float | None = None) -> Path:
    """Writes one row per sample: elapsed_s (relative to first sample),
    current_ma, and (if provided) an instantaneous mW estimate using the
    single end-of-window bus voltage reading (same assumption the energy
    integration uses -- see tests.py). Filename is timestamped to the
    second, with a slug of the test name, so concurrent/rapid tests never
    collide.
    """
    slug = "".join(c if c.isalnum() or c in "-_" else "_" for c in test_name.lower())
    ts = time.strftime("%Y%m%d_%H%M%S")
    path = cache_dir() / f"{ts}_{slug}.csv"

    t0 = samples[0].t if samples else 0.0
    with path.open("w", newline="") as f:
        writer = csv.writer(f)
        header = ["sample_index", "elapsed_s", "current_ma"]
        if bus_voltage_v is not None:
            header.append("power_mw")
        writer.writerow(header)
        for i, s in enumerate(samples):
            row = [i, f"{s.t - t0:.6f}", f"{s.current_ma:.3f}"]
            if bus_voltage_v is not None:
                row.append(f"{s.current_ma * bus_voltage_v:.3f}")
            writer.writerow(row)

    return path


def write_sysinfo_csv(test_name: str, sys_samples: list) -> Path | None:
    """Writes one row per system-resource snapshot (SysSnapshot from
    sysinfo.py) captured during a test: wall-clock time, elapsed_s (relative
    to the first snapshot), CPU%, GPU%, RAM used/total/%, swap used/total/%,
    fan RPM/duty, and die temperature. Returns None (writes nothing) if no
    snapshots were captured (e.g. jtop unavailable for the whole test).
    Filename mirrors write_samples_csv's convention with a "_sysinfo" suffix
    so the two CSVs for one test run are easy to pair up by timestamp+name.
    """
    if not sys_samples:
        return None
    slug = "".join(c if c.isalnum() or c in "-_" else "_" for c in test_name.lower())
    ts = time.strftime("%Y%m%d_%H%M%S")
    path = cache_dir() / f"{ts}_{slug}_sysinfo.csv"

    t0 = sys_samples[0].t
    with path.open("w", newline="") as f:
        writer = csv.writer(f)
        writer.writerow(
            [
                "sample_index",
                "elapsed_s",
                "wall_time",
                "cpu_percent",
                "gpu_percent",
                "ram_used_mb",
                "ram_total_mb",
                "ram_percent",
                "swap_used_mb",
                "swap_total_mb",
                "swap_percent",
                "fan_rpm",
                "fan_percent",
                "temp_c",
            ]
        )
        for i, s in enumerate(sys_samples):
            writer.writerow(
                [
                    i,
                    f"{s.t - t0:.3f}",
                    f"{s.t:.3f}",
                    f"{s.cpu_percent:.2f}",
                    f"{s.gpu_percent:.2f}",
                    f"{s.ram_used_mb:.1f}",
                    f"{s.ram_total_mb:.1f}",
                    f"{s.ram_percent:.2f}",
                    f"{s.swap_used_mb:.1f}",
                    f"{s.swap_total_mb:.1f}",
                    f"{s.swap_percent:.2f}",
                    f"{s.fan_rpm:.1f}" if s.fan_rpm is not None else "",
                    f"{s.fan_percent:.1f}" if s.fan_percent is not None else "",
                    f"{s.temp_c:.2f}" if s.temp_c is not None else "",
                ]
            )

    return path

File: src/csv_log.c
/* csv_log.c -- see csv_log.h */
#include "csv_log.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

static void slugify(const char *name, char *out, size_t out_size) {
    size_t j = 0;
    for (size_t i = 0; name[i] && j + 1 < out_size; i++) {
        char c = name[i];
        char lc = (c >= 'A' && c <= 'Z') ? (char)(c - 'A' + 'a') : c;
        int keep = (lc >= 'a' && lc <= 'z') || (lc >= '0' && lc <= '9') || lc == '-' || lc == '_';
        out[j++] = keep ? lc : '_';
    }
    out[j] = '\0';
}

/* Builds "<cache_dir>/<timestamp>_<slug><suffix>.csv" into `out`. Returns
 * 0 on success, -1 if the cache dir couldn't be determined/created. */
static int build_csv_path(const char *test_name, const char *suffix, char *out, size_t out_size) {
    char *cache_dir = jeu_cache_dir();
    if (!cache_dir) return -1;

    char slug[128];
    slugify(test_name, slug, sizeof(slug));

    time_t now = time(NULL);
    struct tm tm_now;
    localtime_r(&now, &tm_now);
    char ts[32];
    strftime(ts, sizeof(ts), "%Y%m%d_%H%M%S", &tm_now);

    snprintf(out, out_size, "%s/%s_%s%s.csv", cache_dir, ts, slug, suffix);
    free(cache_dir);
    return 0;
}

char *write_samples_csv(const char *test_name, const Sample *samples, size_t n,
                         double bus_voltage_v_or_negative) {
    if (n == 0) return NULL;

    char path[4096];
    if (build_csv_path(test_name, "", path, sizeof(path)) != 0) return NULL;

    FILE *f = fopen(path, "w");
    if (!f) return NULL;

    /* Large stdio buffer so this batched write doesn't get split into many
     * small write() syscalls (same rationale as ina_bench.c). */
    static char buf[1 << 20];
    setvbuf(f, buf, _IOFBF, sizeof(buf));

    int has_power = bus_voltage_v_or_negative >= 0.0;
    if (has_power) {
        fprintf(f, "sample_index,elapsed_s,current_ma,power_mw\n");
    } else {
        fprintf(f, "sample_index,elapsed_s,current_ma\n");
    }

    double t0 = samples[0].t;
    for (size_t i = 0; i < n; i++) {
        double elapsed = samples[i].t - t0;
        if (has_power) {
            fprintf(f, "%zu,%.6f,%.3f,%.3f\n", i, elapsed, samples[i].current_ma,
                    samples[i].current_ma * bus_voltage_v_or_negative);
        } else {
            fprintf(f, "%zu,%.6f,%.3f\n", i, elapsed, samples[i].current_ma);
        }
    }
    fclose(f);
    return strdup(path);
}

char *write_sysinfo_csv(const char *test_name, const SysSnapshot *sys_samples, size_t n) {
    if (n == 0) return NULL;

    char path[4096];
    if (build_csv_path(test_name, "_sysinfo", path, sizeof(path)) != 0) return NULL;

    FILE *f = fopen(path, "w");
    if (!f) return NULL;
    static char buf[1 << 20];
    setvbuf(f, buf, _IOFBF, sizeof(buf));

    fprintf(f,
            "sample_index,elapsed_s,wall_time,cpu_percent,gpu_percent,ram_used_mb,"
            "ram_total_mb,ram_percent,swap_used_mb,swap_total_mb,swap_percent,fan_rpm,"
            "fan_percent,temp_c\n");

    double t0 = sys_samples[0].t;
    for (size_t i = 0; i < n; i++) {
        const SysSnapshot *s = &sys_samples[i];
        fprintf(f, "%zu,%.3f,%.3f,%.2f,%.2f,%.1f,%.1f,%.2f,%.1f,%.1f,%.2f,", i, s->t - t0, s->t,
                s->cpu_percent, s->gpu_percent, s->ram_used_mb, s->ram_total_mb, s->ram_percent,
                s->swap_used_mb, s->swap_total_mb, s->swap_percent);
        if (s->fan_rpm >= 0)
            fprintf(f, "%.1f,", s->fan_rpm);
        else
            fprintf(f, ",");
        if (s->fan_percent >= 0)
            fprintf(f, "%.1f,", s->fan_percent);
        else
            fprintf(f, ",");
        if (s->temp_c > -999.0)
            fprintf(f, "%.2f\n", s->temp_c);
        else
            fprintf(f, "\n");
    }
    fclose(f);
    return strdup(path);
}
